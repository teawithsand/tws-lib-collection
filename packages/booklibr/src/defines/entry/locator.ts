/**
 * Describes the methods by which an Address Book (ABook) entry can be located.
 */
export enum ABookEntryLocatorType {
	/**
	 * The entry is located within the internal object storage system,
	 * accessible via a unique identifier.
	 */
	INTERNAL_OBJECT_STORAGE = 1,

	/**
	 * The entry is located by an external URL.
	 */
	URL = 2,
}

/**
 * Represents a locator for an Address Book entry, specifying how to find it.
 */
export type ABookEntryLocator =
	| {
			/**
			 * Specifies that the ABook entry is located in internal object storage.
			 */
			type: ABookEntryLocatorType.INTERNAL_OBJECT_STORAGE

			/**
			 * Unique identifier of the entry within the internal storage.
			 */
			id: string
	  }
	| {
			/**
			 * Specifies that the ABook entry is located via an external URL.
			 */
			type: ABookEntryLocatorType.URL

			/**
			 * The URL where the entry can be accessed.
			 */
			url: string
	  }
