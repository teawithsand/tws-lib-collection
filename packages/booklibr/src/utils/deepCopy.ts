/**
 * Performs a deep copy of the provided value
 * @param value The value to deep copy
 * @returns A deep copy of the provided value
 */
export const deepCopy = <T>(value: T): T => {
	// Handle null or undefined
	if (value === null || value === undefined) {
		return value;
	}

	// Handle primitive types
	if (typeof value !== 'object') {
		return value;
	}

	// Handle Date objects
	if (value instanceof Date) {
		return new Date(value.getTime()) as unknown as T;
	}

	// Handle Arrays
	if (Array.isArray(value)) {
		return value.map(item => deepCopy(item)) as unknown as T;
	}

	// Handle Objects
	const result: Record<string, unknown> = {};
	Object.entries(value as Record<string, unknown>).forEach(([key, val]) => {
		result[key] = deepCopy(val);
	});

	return result as T;
};