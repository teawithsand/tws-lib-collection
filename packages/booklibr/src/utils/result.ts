export enum ResultType {
	OK = 1,
	ERROR = 2,
}

export type Result<T, E = any> =
	| {
			type: ResultType.OK
			value: T
			error?: undefined
	  }
	| {
			type: ResultType.ERROR
			error: E
			value?: undefined
	  }
