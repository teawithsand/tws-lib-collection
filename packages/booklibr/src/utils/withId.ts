export type Id = string

export type WithId<T> = {
	data: T
	id: string
}
