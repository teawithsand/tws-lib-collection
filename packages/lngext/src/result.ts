/**
 * Enum representing the possible types of a Result
 */
export enum ResultType {
	/**
	 * Represents a successful result
	 */
	Ok = "ok",

	/**
	 * Represents a failed result
	 */
	Error = "error",
}

/**
 * A tagged union type representing either a successful result or an error
 */
export type ResultVariant<T, E = Error> =
	| { type: ResultType.Ok; value: T }
	| { type: ResultType.Error; error: E }

/**
 * Result monad that contains either a value of type T or an error of type E.
 */
export class Result<T, E = Error> {
	private constructor(
		private readonly value: T | undefined,
		private readonly error: E | undefined,
		private readonly isOk: boolean,
	) {}

	/**
	 * Creates a successful Result containing the given value.
	 */
	public static readonly ok = <T, E = Error>(value: T): Result<T, E> => {
		return new Result<T, E>(value, undefined, true)
	}

	/**
	 * Creates a failed Result containing the given error.
	 */
	public static readonly error = <T, E = Error>(error: E): Result<T, E> => {
		return new Result<T, E>(undefined, error, false)
	}

	/**
	 * Creates a Result from a ResultVariant.
	 */
	public static readonly fromType = <T, E = Error>(
		resultType: ResultVariant<T, E>,
	): Result<T, E> => {
		if (resultType.type === ResultType.Ok) {
			return Result.ok<T, E>(resultType.value)
		}
		return Result.error<T, E>(resultType.error)
	}

	/**
	 * Returns true if the Result is successful.
	 */
	public readonly isSuccess = (): boolean => {
		return this.isOk
	}

	/**
	 * Returns true if the Result is a failure.
	 */
	public readonly isFailure = (): boolean => {
		return !this.isOk
	}

	/**
	 * Returns the value if the Result is successful, or throws the error if it's a failure.
	 * @throws E if the Result is a failure
	 */
	public readonly unwrap = (): T => {
		if (!this.isOk) {
			throw this.error
		}
		return this.value as T
	}

	/**
	 * Returns the value if the Result is successful, or the provided default value if it's a failure.
	 */
	public readonly unwrapOr = (defaultValue: T): T => {
		return this.isOk ? (this.value as T) : defaultValue
	}

	/**
	 * Returns the value if the Result is successful, or the result of calling the fallback function with the error.
	 */
	public readonly unwrapOrElse = (fallback: (error: E) => T): T => {
		return this.isOk ? (this.value as T) : fallback(this.error as E)
	}

	/**
	 * Returns the error if the Result is a failure, or throws an error if it's successful.
	 * @throws Error if the Result is successful
	 */
	public readonly unwrapError = (): E => {
		if (this.isOk) {
			throw this.error
		}
		return this.error as E
	}

	/**
	 * Maps a Result<T, E> to Result<U, E> by applying the function to the contained value.
	 */
	public readonly map = <U>(fn: (value: T) => U): Result<U, E> => {
		if (this.isOk) {
			return Result.ok<U, E>(fn(this.value as T))
		}
		return Result.error<U, E>(this.error as E)
	}

	/**
	 * Maps a Result<T, E> to Result<T, F> by applying the function to the contained error.
	 */
	public readonly mapError = <F>(fn: (error: E) => F): Result<T, F> => {
		if (this.isOk) {
			return Result.ok<T, F>(this.value as T)
		}
		return Result.error<T, F>(fn(this.error as E))
	}

	/**
	 * Returns the provided result if this result is successful, otherwise returns the error.
	 */
	public readonly andThen = <U>(
		fn: (value: T) => Result<U, E>,
	): Result<U, E> => {
		if (this.isOk) {
			return fn(this.value as T)
		}
		return Result.error<U, E>(this.error as E)
	}

	/**
	 * Returns this result if it is a failure, otherwise evaluates the function and returns its result.
	 */
	public readonly orElse = (fn: (error: E) => Result<T, E>): Result<T, E> => {
		if (this.isOk) {
			return this
		}
		return fn(this.error as E)
	}

	/**
	 * Returns the ResultType of this Result.
	 */
	public get type(): ResultType {
		return this.isOk ? ResultType.Ok : ResultType.Error
	}

	/**
	 * Converts the Result to a tagged union type for exhaustive pattern matching in switch statements
	 */
	public readonly toType = (): ResultVariant<T, E> => {
		if (this.isOk) {
			return { type: ResultType.Ok, value: this.value as T }
		}
		return { type: ResultType.Error, error: this.error as E }
	}

	/**
	 * Executes the appropriate callback based on whether the Result is successful or a failure.
	 */
	public readonly match = <U>(options: {
		ok: (value: T) => U
		error: (error: E) => U
	}): U => {
		if (this.isOk) {
			return options.ok(this.value as T)
		}
		return options.error(this.error as E)
	}
}
