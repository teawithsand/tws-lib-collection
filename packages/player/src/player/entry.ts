export enum PlayerEntryType {
	BLOB = "blob",
	URL = "url",
}

export type PlayerEntry =
	| { type: PlayerEntryType.BLOB; blob: Blob }
	| { type: PlayerEntryType.URL; url: string }
