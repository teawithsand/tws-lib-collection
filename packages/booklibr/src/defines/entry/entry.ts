import { Result } from "../../utils/result"
import { SerializedError } from "../misc/error"
import { ABookEntryLocator } from "./locator"

export type ABookEntryUploadMetadata = {
	fileName: string | null
	lastModifiedTimestamp: number | null
	fileSize: number
	path: string | null
}

export enum ABookEntryDisposition {
	AUDIO = 1,
	IMAGE = 2,
	DESCRIPTION = 3,
}

export type ABookEntryAudioMetadata = {
	duration: Result<number, SerializedError>
}

export type ABookEntryMetadata = {
	uploadMetadata: ABookEntryUploadMetadata | null
	audioMetadata: ABookEntryAudioMetadata | null
}

export type ABookEntry = {
	disposition: ABookEntryDisposition
	metadata: ABookEntryMetadata
	locator: ABookEntryLocator
}
