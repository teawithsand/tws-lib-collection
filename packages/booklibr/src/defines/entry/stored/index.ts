import { z } from "zod"
import { ResultType } from "../../../utils/result"
import { ABookEntry, ABookEntryDisposition } from "../entry"

// Stored enum copies
export enum StoredABookEntryDispositionV1 {
	AUDIO = 1,
	IMAGE = 2,
	DESCRIPTION = 3,
}

export type StoredABookEntryDisposition = StoredABookEntryDispositionV1

// Stored types and schemas for ABookEntry only
export const StoredABookEntryV1Schema = z.object({
	disposition: z.nativeEnum(StoredABookEntryDispositionV1),
	metadata: z.object({
		uploadMetadata: z
			.object({
				fileName: z.string().nullable(),
				lastModifiedTimestamp: z.number().nullable(),
				fileSize: z.number(),
				path: z.string().nullable(),
			})
			.nullable(),
		audioMetadata: z
			.object({
				duration: z.object({
					type: z.literal("ok").or(z.literal("err")),
					value: z.number().or(z.any()),
				}),
			})
			.nullable(),
	}),
	locator: z.union([
		z.object({
			type: z.literal(1),
			id: z.string(),
		}),
		z.object({
			type: z.literal(2),
			url: z.string(),
		}),
	]),
})
export type StoredABookEntryV1 = z.infer<typeof StoredABookEntryV1Schema>
export type StoredABookEntry = StoredABookEntryV1
export const StoredABookEntrySchema = StoredABookEntryV1Schema

export class ABookEntrySerializer {
	private constructor() {}

	private static readonly toStoredDisposition = (
		d: ABookEntryDisposition,
	): StoredABookEntryDispositionV1 => {
		switch (d) {
			case ABookEntryDisposition.AUDIO:
				return StoredABookEntryDispositionV1.AUDIO
			case ABookEntryDisposition.IMAGE:
				return StoredABookEntryDispositionV1.IMAGE
			case ABookEntryDisposition.DESCRIPTION:
				return StoredABookEntryDispositionV1.DESCRIPTION
			default:
				throw new Error("Unknown ABookEntryDisposition")
		}
	}

	private static readonly fromStoredDisposition = (
		d: StoredABookEntryDispositionV1,
	): ABookEntryDisposition => {
		switch (d) {
			case StoredABookEntryDispositionV1.AUDIO:
				return ABookEntryDisposition.AUDIO
			case StoredABookEntryDispositionV1.IMAGE:
				return ABookEntryDisposition.IMAGE
			case StoredABookEntryDispositionV1.DESCRIPTION:
				return ABookEntryDisposition.DESCRIPTION
			default:
				throw new Error("Unknown StoredABookEntryDispositionV1")
		}
	}

	private static readonly toStoredLocator = (
		locator: ABookEntry["locator"],
	): StoredABookEntry["locator"] => {
		if (locator.type === 1) {
			return { type: 1, id: locator.id }
		}
		return { type: 2, url: locator.url }
	}

	private static readonly fromStoredLocator = (
		locator: StoredABookEntry["locator"],
	): ABookEntry["locator"] => {
		if (locator.type === 1) {
			return { type: 1, id: locator.id }
		}
		return { type: 2, url: locator.url }
	}

	public static readonly serialize = (
		entry: ABookEntry,
	): StoredABookEntry => ({
		disposition: ABookEntrySerializer.toStoredDisposition(
			entry.disposition,
		),
		metadata: entry.metadata
			? {
					uploadMetadata: entry.metadata.uploadMetadata
						? { ...entry.metadata.uploadMetadata }
						: null,
					audioMetadata: entry.metadata.audioMetadata
						? {
								duration:
									entry.metadata.audioMetadata.duration
										.type === ResultType.OK
										? {
												type: "ok",
												value: entry.metadata
													.audioMetadata.duration
													.value,
											}
										: {
												type: "err",
												value: entry.metadata
													.audioMetadata.duration
													.error,
											},
							}
						: null,
				}
			: { uploadMetadata: null, audioMetadata: null },
		locator: ABookEntrySerializer.toStoredLocator(entry.locator),
	})

	public static readonly deserialize = (
		stored: StoredABookEntry,
	): ABookEntry => ({
		disposition: ABookEntrySerializer.fromStoredDisposition(
			stored.disposition,
		),
		metadata: stored.metadata
			? {
					uploadMetadata: stored.metadata.uploadMetadata
						? { ...stored.metadata.uploadMetadata }
						: null,
					audioMetadata: stored.metadata.audioMetadata
						? {
								duration:
									stored.metadata.audioMetadata.duration
										.type === "ok"
										? {
												type: ResultType.OK,
												value: stored.metadata
													.audioMetadata.duration
													.value,
											}
										: {
												type: ResultType.ERROR,
												error: stored.metadata
													.audioMetadata.duration
													.value,
											},
							}
						: null,
				}
			: { uploadMetadata: null, audioMetadata: null },
		locator: ABookEntrySerializer.fromStoredLocator(stored.locator),
	})
}
