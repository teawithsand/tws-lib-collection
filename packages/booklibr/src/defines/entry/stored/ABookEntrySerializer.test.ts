import { describe, expect, test } from "vitest"
import { ResultType } from "../../../utils/result"
import { ABookEntry, ABookEntryDisposition } from "../entry"
import { ABookEntrySerializer, StoredABookEntryDispositionV1 } from "./index"

const makeEntry = (opts: Partial<ABookEntry> = {}): ABookEntry => {
	const {
		disposition = ABookEntryDisposition.AUDIO,
		metadata = {
			uploadMetadata: {
				fileName: "file.mp3",
				lastModifiedTimestamp: 123,
				fileSize: 456,
				path: "/tmp/file.mp3",
			},
			audioMetadata: {
				duration: { type: ResultType.OK, value: 42 },
			},
		},
		locator = { type: 1, id: "abc" },
	} = opts
	return {
		disposition,
		metadata,
		locator,
	}
}

const makeSerializedError = () => ({
	name: "Error",
	message: "fail",
	stack: "stacktrace",
})

describe("ABookEntrySerializer", () => {
	test("serialize and deserialize roundtrip for AUDIO", () => {
		const entry = makeEntry({ disposition: ABookEntryDisposition.AUDIO })
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.disposition).toBe(StoredABookEntryDispositionV1.AUDIO)
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual(entry)
	})

	test("serialize and deserialize roundtrip for IMAGE", () => {
		const entry = makeEntry({ disposition: ABookEntryDisposition.IMAGE })
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.disposition).toBe(StoredABookEntryDispositionV1.IMAGE)
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual(entry)
	})

	test("serialize and deserialize roundtrip for DESCRIPTION", () => {
		const entry = makeEntry({
			disposition: ABookEntryDisposition.DESCRIPTION,
		})
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.disposition).toBe(
			StoredABookEntryDispositionV1.DESCRIPTION,
		)
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual(entry)
	})

	test("serialize handles null metadata fields", () => {
		const entry = makeEntry({
			metadata: {
				uploadMetadata: null,
				audioMetadata: null,
			},
		})
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.metadata.uploadMetadata).toBeNull()
		expect(stored.metadata.audioMetadata).toBeNull()
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual(entry)
	})

	test("serialize handles missing metadata", () => {
		const entry = makeEntry({
			metadata: { uploadMetadata: null, audioMetadata: null },
		})
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.metadata.uploadMetadata).toBeNull()
		expect(stored.metadata.audioMetadata).toBeNull()
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual({
			...entry,
			metadata: { uploadMetadata: null, audioMetadata: null },
		})
	})

	test("serialize handles locator type 2", () => {
		const entry = makeEntry({
			locator: { type: 2, url: "http://example.com" },
		})
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.locator).toEqual({ type: 2, url: "http://example.com" })
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual(entry)
	})

	test("serialize handles audioMetadata duration error", () => {
		const entry = makeEntry({
			metadata: {
				uploadMetadata: null,
				audioMetadata: {
					duration: {
						type: ResultType.ERROR,
						error: makeSerializedError(),
					},
				},
			},
		})
		const stored = ABookEntrySerializer.serialize(entry)
		expect(stored.metadata.audioMetadata?.duration.type).toBe("err")
		expect(stored.metadata.audioMetadata?.duration.value).toEqual(
			makeSerializedError(),
		)
		const roundtrip = ABookEntrySerializer.deserialize(stored)
		expect(roundtrip).toEqual(entry)
	})

	test("deserialize produces stable snapshots for all dispositions", () => {
		// Inline serialized versions for each disposition with correct types
		const audioSerialized = {
			disposition: StoredABookEntryDispositionV1.AUDIO,
			metadata: {
				uploadMetadata: {
					fileName: "file.mp3",
					lastModifiedTimestamp: 123,
					fileSize: 456,
					path: "/tmp/file.mp3",
				},
				audioMetadata: {
					duration: { type: "ok" as const, value: 42 },
				},
			},
			locator: { type: 1 as const, id: "abc" },
		}
		const imageSerialized = {
			disposition: StoredABookEntryDispositionV1.IMAGE,
			metadata: {
				uploadMetadata: {
					fileName: "file.mp3",
					lastModifiedTimestamp: 123,
					fileSize: 456,
					path: "/tmp/file.mp3",
				},
				audioMetadata: {
					duration: { type: "ok" as const, value: 42 },
				},
			},
			locator: { type: 1 as const, id: "abc" },
		}
		const descriptionSerialized = {
			disposition: StoredABookEntryDispositionV1.DESCRIPTION,
			metadata: {
				uploadMetadata: {
					fileName: "file.mp3",
					lastModifiedTimestamp: 123,
					fileSize: 456,
					path: "/tmp/file.mp3",
				},
				audioMetadata: {
					duration: { type: "ok" as const, value: 42 },
				},
			},
			locator: { type: 1 as const, id: "abc" },
		}

		expect(
			ABookEntrySerializer.deserialize(audioSerialized),
		).toMatchSnapshot()
		expect(
			ABookEntrySerializer.deserialize(imageSerialized),
		).toMatchSnapshot()
		expect(
			ABookEntrySerializer.deserialize(descriptionSerialized),
		).toMatchSnapshot()
	})
})
