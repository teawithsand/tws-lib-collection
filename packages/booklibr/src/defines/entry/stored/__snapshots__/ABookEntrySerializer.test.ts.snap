// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ABookEntrySerializer > deserialize produces stable snapshots for all dispositions 1`] = `
{
  "disposition": 1,
  "locator": {
    "id": "abc",
    "type": 1,
  },
  "metadata": {
    "audioMetadata": {
      "duration": {
        "type": 1,
        "value": 42,
      },
    },
    "uploadMetadata": {
      "fileName": "file.mp3",
      "fileSize": 456,
      "lastModifiedTimestamp": 123,
      "path": "/tmp/file.mp3",
    },
  },
}
`;

exports[`ABookEntrySerializer > deserialize produces stable snapshots for all dispositions 2`] = `
{
  "disposition": 2,
  "locator": {
    "id": "abc",
    "type": 1,
  },
  "metadata": {
    "audioMetadata": {
      "duration": {
        "type": 1,
        "value": 42,
      },
    },
    "uploadMetadata": {
      "fileName": "file.mp3",
      "fileSize": 456,
      "lastModifiedTimestamp": 123,
      "path": "/tmp/file.mp3",
    },
  },
}
`;

exports[`ABookEntrySerializer > deserialize produces stable snapshots for all dispositions 3`] = `
{
  "disposition": 3,
  "locator": {
    "id": "abc",
    "type": 1,
  },
  "metadata": {
    "audioMetadata": {
      "duration": {
        "type": 1,
        "value": 42,
      },
    },
    "uploadMetadata": {
      "fileName": "file.mp3",
      "fileSize": 456,
      "lastModifiedTimestamp": 123,
      "path": "/tmp/file.mp3",
    },
  },
}
`;
